javascript/computer_science/project_hash_map.md

```diff
@@ -0,0 +1,227 @@
+
+### Computation complexity
+
+A hash map is very efficient in its insertion, retrieving and removing operations. That is because we use array indexes to do those operations. A hash map has an average case complexity of `O(1)` for the following methods:
+- Insertion
+- Retrieving
+- Deletion
+
+Assuming we have a good hash map written. The worst case of those operations would be `O(n)` and that happens when we have all our data hashes to the same exact bucket.
+
+The Growth of our hash map has the complexity of `O(n)` at all times.
+
+### Assignment
+
+<div class="lesson-content__panel" markdown="1">
+
+#### Self restriction and limitation
+
+  Before we get started, we need to lay down some ground rules. As we all know JavaScript is a dynamic language the restriction we have in other languages does not exist in JavaScript for example if you create an array of size `16` to be your buckets size, nothing stopping you from storing items at index `500`. This defeats the purpose we are trying to demonstrate, so we need to put some self restriction to work around this.
+
+  1. When you create a function that takes a hash code value and returns the bucket index, use this snippet before the function returns to make sure it's accessing inbound array indexes only.
+
+      ```javascript
+      if (index < 0 || index >= buckets.length) {
+        throw new Error("Trying to access index out of bound");
+      }
+      ```
+
+  1. Inside the class DO NOT use Objects to store your key value pairs, And DO NOT use the array method `push` to insert your nodes into the buckets, always use indexes.
+
+#### Time to create a real hash map
+
+  1. Start by creating a HashMap class or factory function, up to you. Then proceed to create the following methods:
+
+  1. `hash` takes a value and produces a hash code with it. There are a lot of thought that goes into creating the most optimal hashing function, it is up to you if you want to do a research to find a good hashing function but do not dive too deep into the rabbit hole. Make sure to test that it's not producing too many similar hash codes that will lead to too many collisions.
+
+  1. `set` takes two arguments, the first is a key and the second is a value that is assigned to this key.
+
+     - Remember to grow your buckets size when it needs to, by calculating the `load factor` and `capacity`.
+
+  1. `get` takes one argument as a key and returns the value that is assigned to this key.
+
+  1. `has` takes a key as an argument and checks if your hash map has that key.
+
+  1. `remove` takes a key as argument and remove it from the hash table.
```

Comment:
Should `remove()` return anything? Usually it does return something, if for no other reason than to let the developer know if an item has been deleted or not.

[JavaScript Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/delete#return_value) returns `true` if the key was deleted and `false` if it wasn't.

In Ruby hashes, [the value for the deleted item is returned](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/delete#return_value) (or `nil` if the key wasn't found).

Either one of those approaches could make sense to me. Maybe with this JS project, we can design it like the JS `Map`?

Reply:
This is true, I don't think I have come across a language that returns nothing with its remove function. Good catch, I think returning true or false will suffice for the JS implementation, maybe it can be reworked in Ruby lesson to be more similar to the hash map in ruby where it returns `nil` if not found? That part is up to you